### PR TITLE
Add locking to Hive registration

### PR DESCRIPTION
**Pull request by zliu41**

### CODE REVIEW

**Comment by zliu41** [comment]
@sahilTakiar @ibuenros

File: gobblin-utility/src/main/java/gobblin/util/AutoCloseableLock.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+  private final Lock lock;
+
+  @Override
+  public void close() {
+    this.lock.unlock();
+  }
+
+  @Override
+  public void lock() {
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/AutoCloseableLock.java]: Are you calling this method at any point?

File: gobblin-utility/src/main/java/gobblin/util/AutoCloseableLock.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.concurrent.locks.Lock;
+
+import lombok.AllArgsConstructor;
+
+
+/**
+ * An auto-closeable {@link Lock} that can be used with try-with-resources.
+ */
+@AllArgsConstructor
+public class AutoCloseableLock implements Lock, AutoCloseable {
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/AutoCloseableLock.java]: I would prefer if the lock automatically locked on construction, and to make everything private but `close()`. That way, the usage must always be:
```java
try (new AutoCloseableLock(lock)) {
...
}
```

**Comment by zliu41** [comment]
@ibuenros updated d8ba8ad

File: gobblin-hive-registration/src/main/java/gobblin/hive/HivePartition.java [hunk 3]
```diff
@@ -45,6 +45,11 @@ private HivePartition(Builder builder) {
     this.values = ImmutableList.<String> copyOf(builder.values);
   }
 
+  @Override
+  public String toString() {
+    return this.values.toString();
```
Comment (ibuenros) [inline gobblin-hive-registration/src/main/java/gobblin/hive/HivePartition.java]: You probably want to use `ArrayUtils.toString(list.toArray())` (as it is, you will get the hashCode instead).

File: gobblin-utility/src/main/java/gobblin/util/AutoCloseableLock.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package gobblin.util;
+
+import java.util.concurrent.locks.Lock;
+
+
+/**
+ * An auto-closeable {@link Lock} that can be used with try-with-resources.
+ * The lock is locked in the constructor and unlocked in {@link #close()}.
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/AutoCloseableLock.java]: Can you include usage in the javadoc?

File: gobblin-hive-registration/src/main/java/gobblin/hive/HivePartition.java (hunk 3 above)
Comment (zliu41) [inline gobblin-hive-registration/src/main/java/gobblin/hive/HivePartition.java]: `Collection.toString` prints all elements in the collection.

**Comment by zliu41** [comment]
@ibuenros updated

**Comment by ibuenros** [comment]
LGTM

**Pull request merged**